### PR TITLE
[API] use softmax with length, and interleaved matmul for BERT

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
This PR changes the input layout of BERTEncoder from NTC to TNC, so that we can adopt the fast fused self attention op introduced by @Caenorst  and @TaoLv . 

For BERTModel API, the input layout remains unchanged. If users obtain the BERT model via the get_model API, they don't need to make any code change to run the optimized version (other than upgrading the gluon-nlp version). 

The tests won't pass as we need to wait for MXNet's nightly build since Jan 4th, otherwise the CPU op is missing. 

On p3.16xlarge, BERT base, seq_len=512, batch_size=256
- latency/batch before this commit: 611ms
- latency/batch after this commit: 386ms

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@leezu I had a temporary fix due to https://github.com/apache/incubator-mxnet/issues/17220 
will that be fixed with gluon 2.0 design?

**Comment by leezu** [comment]
@eric-haibin-lin yes, https://github.com/apache/incubator-mxnet/issues/17220#issuecomment-571088223

**Comment by eric-haibin-lin** [comment]
@muhyun this helps GPT-2, too

**Comment by sxjscience** [comment]
@eric-haibin-lin Will we achieve similar speed up if we fuse the kernel + using `NTC` layout?

**Comment by leezu** [comment]
Merged master due to https://github.com/dmlc/gluon-nlp/pull/1096. Should be possible to merge now

**Comment by mli** [comment]
Job PR-1091/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1091/5/index.html

**Comment by leezu** [comment]
@eric-haibin-lin 

```
[2020-01-15T17:33:28.776Z] mxnet.base.MXNetError: Error in operator bertencoder0_transformer0_dotproductselfattentioncell0_interleaved_matmul_selfatt_valatt0: [17:21:45] src/operator/numpy/linalg/./../../tensor/../elemwise_op_common.h:135: Check failed: assign(&dattr, vec.at(i)): Incompatible attr in node bertencoder0_transformer0_dotproductselfattentioncell0_interleaved_matmul_selfatt_valatt0 at 1-th input: expected float16, got float32
```

**Comment by mli** [comment]
Job PR-1091/7 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1091/7/index.html

**Comment by mli** [comment]
Job PR-1091/9 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1091/9/index.html

**Comment by eric-haibin-lin** [comment]
@leezu could you help build the latest mxnet 1.6.0.rc1 for our CI pipeline? thanks!

**Comment by leezu** [comment]
@eric-haibin-lin done

**Comment by mli** [comment]
Job PR-1091/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1091/10/index.html

**Comment by szhengac** [comment]
Have u tried training the model?

**Comment by eric-haibin-lin** [comment]
Yes. The one I am running contains the new dataset loader and this change. So far the loss looks normal

File: src/gluonnlp/model/bert.py [hunk 1]
```diff
@@ -30,13 +30,212 @@
...
+
+    def _collect_params_with_prefix(self, prefix=''):
+        # the registered parameter names in v0.8 are the following:
+        # prefix_proj_query.weight, prefix_proj_query.bias
+        # prefix_proj_value.weight, prefix_proj_value.bias
+        # prefix_proj_key.weight, prefix_proj_key.bias
+        # this is a temporary fix to keep backward compatibility, due to an issue in MXNet:
+        # https://github.com/apache/incubator-mxnet/issues/17220
+        if prefix:
+            prefix += '.'
```
Comment (szhengac) [inline src/gluonnlp/model/bert.py]: Doc says `prefix_` but u add `.`?

File: src/gluonnlp/model/bert.py [hunk 2]
```diff
@@ -30,13 +30,212 @@
...
+        ret = {prefix + 'proj_' + k.replace('_', '.') : v for k, v in self._reg_params.items()}
+        for name, child in self._children.items():
+            ret.update(child._collect_params_with_prefix(prefix + name))
+        return ret
+
+    # pylint: disable=arguments-differ
+    def hybrid_forward(self, F, qkv, valid_len, query_bias, key_bias, value_bias,
+                       query_weight, key_weight, value_weight):
+        in_bias = F.concat(query_bias, key_bias, value_bias, dim=0)
```
Comment (szhengac) [inline src/gluonnlp/model/bert.py]: Since u concat everything, why don't we just create a large bias and weight?

File: src/gluonnlp/model/bert.py [hunk 3]
```diff
@@ -30,13 +30,212 @@
...
+        - **inputs** : input sequence. Shape (length, batch_size, C_in)
+        - **valid_length** : valid length of inputs for attention. Shape (batch_size, length)
+
+    Outputs:
+        - **outputs**: output tensor of the transformer encoder cell.
+            Shape (length, batch_size, C_out)
+        - **additional_outputs**: the additional output of all the BERT encoder cell.
+    """
+    def __init__(self, units=128, hidden_size=512, num_heads=4,
+                 dropout=0.0, output_attention=False,
```
Comment (szhengac) [inline src/gluonnlp/model/bert.py]: `use_bias : bool` is missing.

File: src/gluonnlp/model/bert.py (hunk 2 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/bert.py]: that leads to different optimizer.step

File: src/gluonnlp/model/bert.py (hunk 3 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/bert.py]: We didn't have this flag in the previous version. https://github.com/dmlc/gluon-nlp/blob/v0.8.x/src/gluonnlp/model/bert.py#L198 
Is it necessary to add this flag?

File: src/gluonnlp/model/bert.py (hunk 1 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/bert.py]: this follows the original implementation: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/block.py#L409-L415
We only change the part after `proj`

File: src/gluonnlp/model/bert.py (hunk 3 above)
Comment (szhengac) [reply src/gluonnlp/model/bert.py]: As attention has its flag, so I think it is better to add it.

File: src/gluonnlp/model/bert.py (hunk 2 above)
Comment (szhengac) [reply src/gluonnlp/model/bert.py]: that's fair. we can try in the future and see if it does not affect the convergence.

File: src/gluonnlp/model/bert.py (hunk 3 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/model/bert.py]: added

**Comment by mli** [comment]
Job PR-1091/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-1091/11/index.html
